### DISCUSSION TITLE
✨ API for creating Dagger SDKs

### DISCUSSION THREAD

**Discussion by jpadams** [dagger/dagger#12513] [Ideas]
### What are you trying to do?

A consistent, more accessible way to generate new SDKs and add features to be implemented by existing SDKs.

### Why is this important to you?

This will allow for the SDKs that users need to be produced by more community members and make them easier to maintain.

### How are you currently working around this?

SDKs need deeper knowledge of and integration with Dagger core today.

cc @samalba

**Comment by sipsma** [comment]
Passing thought, given the fact that:
1. Quite a bit of the work here is pretty monotonous boilerplate (e.g. the codegen templates)
2. We have a pre-existing set of examples (go, python, nodejs)

This feels like a potentially good use case for LLMs. It's too big an input+output to just plug into chatgpt, but I would not be surprised if a slightly more specialized setup of gpt or similar could take care of a lot of the boilerplate for something like this. Even if the result isn't perfect, if it's pretty good and just requires a few manual tweaks here and there that could go a long way.

I also have no idea what I'm talking about when it comes to whole AI space, so maybe this is not actually viable.

**Comment by kjuulh** [comment]
I see a few issues in general. You kind of want to still hold the reins of what is in the sdks you don't want a super big sprawl because then adding features, and tweaking stuff will become a lot of work.

Each sdk has some core bits that needs to be added to a degree some of it could be solved using a shared library approach, most of the current languages technically could use a dll/so file approach for downloading the engine and creating a connection to it. This would reduce most the actual code in the sdks. You probably also could consider embedding the core library code as wasm, don't know if that is any better, but it is an alternative at least.

The actual querier would again probably have to be specific to each language, as they are reasonably different, and you also want the feel of the language to be natural, and not like a generated api client =D.

For the codegen bits a shared set of utility functions for the introspection result as well as some helper functions would probably be as far as I'd go. Each language again has reasonably different syntax, so I'd leave that functionality to the implementation. 

This is just my 2 cents after building the rust-sdk.

**Comment by d3rp3tt3** [comment]
CC @helderco
